base/compiler/typelimits.jl

```diff
@@ -391,7 +391,19 @@ function tmerge(@nospecialize(typea), @nospecialize(typeb))
                         widen = tuplemerge(unwrap_unionall(ti)::DataType, unwrap_unionall(tj)::DataType)
                         widen = rewrap_unionall(rewrap_unionall(widen, ti), tj)
                     else
-                        widen = typenames[i].wrapper
+                        wr = typenames[i].wrapper
+                        uw = unwrap_unionall(wr)
+                        ui = unwrap_unionall(ti)
+                        uj = unwrap_unionall(tj)
+                        merged = wr
+                        for k = 1:length(uw.parameters)
+                            if ui.parameters[k] === uj.parameters[k]
+                                merged = merged{ui.parameters[k]}
+                            else
+                                merged = merged{uw.parameters[k]}
+                            end
+                        end
+                        widen = rewrap_unionall(merged, wr)
```

Comment:
I wonder if this is entirely sound, due mainly to losing `where` bounds, but also wondering if it'd mismatch TypeVars coming from different places. What if we had something like:
```julia
struct MyType{A, B, C, D} end
T = TypeVar(:T)
S = TypeVar(:S)
A = UnionAll(T, MyType{T, S, T, Pair{T, S}})
B = UnionAll(T, MyType{S, T, T, Pair{T, S}})
```
The merge result would appear to be:
```julia
MyType{A, B, T, Pair{T, S}} where B where A
```

Which, I think the only issue it has is that it's missing `where T`, but just checking

Reply:
You're right. I'll exclude anything with free typevars.